syntax = "proto3";
// import "google/protobuf/any.proto";
// import "google/protobuf/wrappers.proto";
import "google/protobuf/descriptor.proto";

package nanorpc2;

extend google.protobuf.MessageOptions {
	bool event_source = 51224;
}

extend google.protobuf.MethodOptions {
	string property_name = 51223;
	bool async = 51225;
}

enum RpcStatus {
	RpcSucceeded = 0;
	RpcChannelFailure = 1;
	RpcUnknownMethod = 2;
	RpcProtocolError = 3;
	RpcUnknownInterface = 4;
	RpcInvalidCallParameter = 5;
}

message RpcCall {
	string service = 1;
	string method = 2;
	uint32 object_id = 3;
	bytes call_data = 4;
}

message RpcResult {
	RpcStatus status = 1;
	string error_message = 2;
	bytes result_data = 3;
}

message RpcMessage {
	int32 id = 1;
	RpcCall call = 2;
	RpcResult result = 3;
}

message RpcVoid {
}

message RpcObject {
	uint32 object_id = 1;
}

message RpcEvent {
	string event_name = 1;
}